src/compiler/checker.ts

```diff
@@ -15589,6 +15589,11 @@ m2: ${(this.mapper2 as unknown as DebugTypeMapper).__debugToString().split("\n")
                         return accessFlags & AccessFlags.IncludeUndefined ? getUnionType([indexInfo.type, undefinedType]) : indexInfo.type;
                     }
                     errorIfWritingToReadonlyIndex(indexInfo);
+                    if (accessFlags & AccessFlags.IncludeUndefined && objectType.symbol && objectType.symbol.flags & (SymbolFlags.RegularEnum | SymbolFlags.ConstEnum)) {
+                        if (indexType.symbol && indexType.flags & TypeFlags.EnumLiteral && getParentOfSymbol(indexType.symbol) === objectType.symbol) {
+                            return indexInfo.type;
+                        }
+                    }

```

Comment:
Both the initial test `accessFlags & AccessFlags.IncludeUndefined` and one of the possible outcomes `indexInfo.type` are shared between the inserted block and the final `return`. This doesn't need to be a giant ternary but all the logic should be folded into a coherent control flow structure here.

Reply:
Made some changes here to fix the flow.